### ISSUE TITLE
Running custom commands before compiling (with make -j)

### ISSUE THREAD

**Issue by kankaristo** [premake/premake-core#286]
I started using Premake yesterday and I'm hoping to start using it instead of qmake. It's been great so far, thanks for an awesome tool!

I've got most things working, but I'm still having a problem running custom commands before any compiling takes place. I had similar problems with qmake, which is one of the reasons I'm switching to Premake.

One of the custom commands can take over 10 seconds to finish and it *must* be run before compiling, because it generates some of the code (actually, most of the code, nearly 200 000 lines).

So, I'm trying to use Premake to run [SWIG](http://www.swig.org/) to generate code before `make` attempts to compile it.

If I put the commands in a project's `prebuildcommands`:
```lua
project "MyProject"
    prebuildcommands{
        "./swig.sh";
    }
    
    [...]
```

...it works fine if I only use `make -j1` (a single job). This is exactly the same as it was with qmake; they're only "prebuild" for one of the `make` jobs, not the whole build process.

If I put the commands in a separate project and make it a dependency:
```lua
project "MyProject"
    links "SWIG"
    
    [...]

project "SWIG"
    kind "Utility"
    
    prebuildcommands{
        "./swig.sh";
    }
```

This time, the command is run before anything in "MyProject" is compiled (great!), but then I get an error about `g++` having no input files in the "SWIG" project. There's a warning about "Utility" being and unsupported "kind". Is it unsupported for `gmake` or is it completely unimplemented (no mention in the documentation)?

Is there some way to get this to work? I'd hate to resort to the same kind of hacks I had to use with qmake...

**Comment by kankaristo** [comment]
Also, I'm using the release version from Premake's website. I'll try the Git version tomorrow to see if that changes anything.

**Comment by kankaristo** [comment]
I started looking around in the code and figured out how to get this to work. You guys have a fantastic introduction on how to extend Premake, it was really easy to get into.

I'm just getting started with Premake, so I might be doing something evil, but it seems to work well. At least I'm probably not breaking anything, because the "Utility" kind was not supported with gmake before.

Pull request coming soon.

**Comment by kankaristo** [comment]
This is working well enough for me, so I'll close the issue.

**Comment by TurkeyMan** [comment]
I'm having this problem where -j# is causing build ordering issues too. Did you solve this by moving those jobs into a separate utility project? That seems a bit lame. I wonder if there's a better solution possible.
I'd like to think `prebuildcommands` should actually be 'pre build' commands... I'm gonna open this again, and think on it some more. I'll need to sort this out at work.

**Comment by kankaristo** [comment]
Yep, I made a utility project, which the main project depends on. But I agree that `prebuildcommands` should be run before any actual compiling, even with multiple make jobs.

**Comment by na-Itms** [comment]
Is there any news about this issue? We hit it when trying to generate a root source file with [CxxTest](http://cxxtest.com/). In the meantime we are going to use the Utility workaround, but prebuild commands should definitely be hard prerequisites for building the project files.

**Comment by seragh** [comment]
Same issue with gmake2 action, tho the workaround with Utility doesn't work anymore as no TARGET is defined in the resulting Makefile and so will never trigger, luckily I could use a project with "kind Makefile" instead.
